### PR TITLE
feat: added the features sub project in the micro front end way

### LINKED ISSUE #267: Create the features module in the micro front end way
**Is your feature request related to a problem? Please describe.**
Create the features module in the micro front end way

### PR DESCRIPTION
**Pull request by J-DK**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #267

## Type of change

Please check options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
         

Clones added
============
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

File: projects/ff4j-features/e2e/protractor.conf.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  capabilities: {
+    browserName: 'chrome'
+  },
+  directConnect: true,
+  baseUrl: 'http://localhost:4200/',
+  framework: 'jasmine',
+  jasmineNodeOpts: {
+    showColors: true,
+    defaultTimeoutInterval: 30000,
+    print: function() {}
```
Comment (paul58914080) [inline projects/ff4j-features/e2e/protractor.conf.js]: Codacy Issue found: [Expected method shorthand. (object-shorthand)](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416)

File: build-elements/ff4j-features.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const fs = require('fs-extra');
+const concat = require('concat');
+
+(async function build() {
```
Comment (paul58914080) [inline build-elements/ff4j-features.js]: Codacy Issue found: [Move the invocation into the parens that contain the function.](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416)

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
         

Clones added
============
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
- e2e/protractor.conf.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

**Comment by paul58914080** [comment]
Hello @J-DK ,

The build is failing because you are not using chrome headless in karma

**Refer**
- https://github.com/paul58914080/ff4j-web-console/blob/master/karma.conf.js#L33
- https://github.com/paul58914080/ff4j-web-console/blob/master/karma.conf.js#L36-L48

File: package.json [hunk 3]
```diff
@@ -39,25 +40,31 @@
...
     "rxjs": "~6.5.4",
     "tslib": "^1.10.0",
     "zone.js": "~0.10.2"
   },
   "devDependencies": {
-    "@angular-devkit/build-angular": "~0.900.7",
-    "@angular/cli": "~9.0.7",
+    "@angular-builders/custom-webpack": "^9.0.0",
+    "@angular-devkit/build-angular": "~0.901.0",
+    "@angular-builders/custom-webpack": "^9.0.0",
```
Comment (paul58914080) [inline package.json]: Codacy Issue found: [Duplicate field '@angular-builders/custom-webpack'](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416)

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
         

Clones added
============
- e2e/protractor.conf.js  1
- projects/ff4j-features/karma.conf.js  2
- projects/ff4j-features/e2e/protractor.conf.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- projects/ff4j-features/e2e/protractor.conf.js  1
- projects/ff4j-features/karma.conf.js  1
         

Clones added
============
- e2e/protractor.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
- projects/ff4j-features/karma.conf.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- projects/ff4j-features/e2e/protractor.conf.js  1
- projects/ff4j-features/karma.conf.js  1
         

Clones added
============
- projects/ff4j-features/e2e/protractor.conf.js  1
- projects/ff4j-features/karma.conf.js  2
- e2e/protractor.conf.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

File: e2e/src/app.po.ts [hunk 4]
```diff
@@ -6,6 +6,6 @@ export class AppPage {
   }
 
   getTitleText(): Promise<string> {
-    return element(by.css('ff4j-root .content span')).getText() as Promise<string>;
+    return element(by.css('ff4j-root div h1')).getText() as Promise<string>;
```
Comment (paul58914080) [inline e2e/src/app.po.ts]: Thank you for fixing this 👍

File: build-elements/ff4j-features.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  const files = [
+    './dist/ff4j-features/runtime-es5.js',
+    './dist/ff4j-features/polyfills-es5.js',
+    './dist/ff4j-features/main-es5.js'
```
Comment (paul58914080) [inline build-elements/ff4j-features.js]: I think we need to use 2015 as it is a step forward

https://johnpapa.net/es5-es2015-typescript/

File: build-elements/ff4j-features.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  ];
+
+  await fs.ensureDir('elements');
+  await concat(files, 'elements/ff4j-features.js');
+  await fs.copyFile(
+    './dist/ff4j-features/styles.css',
+    'elements/ff4j-features.styles.css'
+  );
+}());
```
Comment (paul58914080) [inline build-elements/ff4j-features.js]: Refer https://github.com/paul58914080/micro-fontend-todo-kata-final/blob/master/projects/create-todo/build/bundle.js#L1-L14

```javascript
const fs = require('fs-extra');
const concat = require('concat');
const del = require('del');

(async function build() {
  const files = [
    'dist/create-todo/runtime-es2015.js',
    'dist/create-todo/polyfills-es2015.js',
    'dist/create-todo/main-es2015.js'
  ];
  await del(['dist/elements/create-todo.js']);
  await fs.ensureDir('dist/elements');
  await concat(files, 'dist/elements/create-todo.js');
})();
```

You would have to delete and then concatenate. Style is not required, I think is is built in

File: projects/ff4j-features/build/extra-webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+module.exports = {
+  output: {
+    jsonpFunction: 'webpackJsonpFf4jFeatures',
+    library: 'ff4jFeatures'
```
Comment (paul58914080) [inline projects/ff4j-features/build/extra-webpack.config.js]: library is not required

File: projects/ff4j-features/tslint.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{
+  "extends": "../../tslint.json",
+  "rules": {
+    "directive-selector": [
+      true,
+      "attribute",
+      "ff4jFeatures",
```
Comment (paul58914080) [inline projects/ff4j-features/tslint.json]: make it just `ff4j`

File: projects/ff4j-features/tslint.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      "camelCase"
+    ],
+    "component-selector": [
+      true,
+      "element",
+      "ff4j-features",
```
Comment (paul58914080) [inline projects/ff4j-features/tslint.json]: make it just `ff4j`

File: build-elements/ff4j-features.js (hunk 6 above)
Comment (J-DK) [reply build-elements/ff4j-features.js]: Yeah, I was aware but is it mandatory? because even without that, it ll first delete and replace I think

**Comment by paul58914080** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- projects/ff4j-features/karma.conf.js  1
- projects/ff4j-features/e2e/protractor.conf.js  1
         

Clones added
============
- projects/ff4j-features/karma.conf.js  2
- projects/ff4j-features/e2e/protractor.conf.js  1
- e2e/protractor.conf.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paul58914080/ff4j-web-console/pullRequest?prid=5248416&bid=17200322)

**Pull request closed without merging**